finality-aleph/src/party/mod.rs

```diff
@@ -254,6 +258,342 @@ where
+            HashSet::from_iter(early_started),
+            "`session_early_started` mismatch"
+        );
+
+        assert_eq!(
+            *controller
+                .node_session_manager
+                .session_stopped
+                .lock()
+                .unwrap(),
+            HashSet::from_iter(stopped),
+            "`session_stopped` mismatch"
+        );
+
+        assert_eq!(
+            *controller
+                .node_session_manager
+                .nonvalidator_session_started
+                .lock()
+                .unwrap(),
+            HashSet::from_iter(non_validator_started),
+            "`nonvalidator_session_started` mismatch"
+        );
+    }
+
+    #[tokio::test(flavor = "multi_thread")]
+    async fn party_starts_session_for_node_in_authorities() {
+        let (party, mut controller) = create_mocked_consensus_party();
+
+        controller
+            .node_session_manager
+            .set_node_id(Some(UintAuthorityId(0).to_public_key()));
+
+        let _party_handle = tokio::spawn(party.run());
+
+        let authorities: Vec<_> = (0..10)
+            .map(|id| UintAuthorityId(id).to_public_key())
+            .collect();
+        set_authorities_for_sessions(
+            &mut controller,
+            vec![
+                (SessionId(0), authorities.clone()),
+                (SessionId(1), authorities),
+            ],
+        )
+        .await;
+
+        sleep(Duration::from_millis(100)).await;
+        simulate_n_blocks(0, SESSION_PERIOD, &controller);
+        sleep(Duration::from_millis(
+            SESSION_STATUS_CHECK_PERIOD.as_millis() as u64 + 100,
+        ))
+        .await;
+
+        assert_session_states(
+            &controller,
+            vec![SessionId(0), SessionId(1)],
+            vec![SessionId(1)],
+            vec![SessionId(0)],
+            vec![],
+        );
+    }
+
+    #[tokio::test(flavor = "multi_thread")]
+    async fn party_run_3_authorities_sessions() {
+        let (party, mut controller) = create_mocked_consensus_party();
+
+        controller
+            .node_session_manager
+            .set_node_id(Some(UintAuthorityId(0).to_public_key()));
+
+        let _party_handle = tokio::spawn(party.run());
+
+        let authorities: Vec<_> = (0..10)
+            .map(|id| UintAuthorityId(id).to_public_key())
+            .collect();
+        set_authorities_for_sessions(
+            &mut controller,
+            vec![
+                (SessionId(0), authorities.clone()),
+                (SessionId(1), authorities.clone()),
+                (SessionId(2), authorities.clone()),
+                (SessionId(3), authorities),
+            ],
+        )
+        .await;
+
+        sleep(Duration::from_millis(100)).await;
+
+        for i in 0..3 {
+            simulate_n_blocks(SESSION_PERIOD * i, SESSION_PERIOD, &controller);
+            sleep(Duration::from_millis(
+                SESSION_STATUS_CHECK_PERIOD.as_millis() as u64 + 100,
+            ))
```

Comment:
do we have to sleep every time?

Reply:
yes to gice time for the party to see that the last block of the session was finalized. I also wanted to test here sequential sessions without having any session skipped because it was too 'fast'